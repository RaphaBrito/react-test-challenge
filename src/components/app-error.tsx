export function AppError() {
  return <h1>Que pena, algo de errado aconteceu :(</h1>;
}
