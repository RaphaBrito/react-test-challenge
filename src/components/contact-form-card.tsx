import { useState } from "react";

import styles from "@/styles/card.module.css";
import { type Contact, type ContactFormData } from "@/types/contact";

export interface ContactFormCardProps {
  contact?: Contact;
  onConfirm: (formData: ContactFormData) => void;
  onCancel: () => void;
}

export function ContactFormCard({
  contact,
  onConfirm,
  onCancel,
}: ContactFormCardProps) {
  const [name, setName] = useState(contact?.name ?? "");
  const [email, setEmail] = useState(contact?.email ?? "");
  const [phone, setPhone] = useState(contact?.phone ?? "");

  const isValid = (): boolean => {
    return (
      name.trim().length !== 0 &&
      email.trim().length !== 0 &&
      phone.trim().length !== 0
    );
  };

  const handleConfirm = () => {
    const formData = { name, email, phone } satisfies ContactFormData;

    onConfirm(formData);
  };

  const handleCancel = onCancel;

  return (
    <div className={styles.card}>
      <input
        value={name}
        onChange={(event) => setName(event.currentTarget.value)}
        placeholder="Name"
      />

      <input
        value={phone}
        onChange={(event) => setPhone(event.currentTarget.value)}
        placeholder="Phone"
      />

      <input
        value={email}
        onChange={(event) => setEmail(event.currentTarget.value)}
        placeholder="Email"
      />

      <div className={styles.cardFooter}>
        <button onClick={handleConfirm} disabled={!isValid()}>
          Salvar
        </button>
        <button onClick={handleCancel}>Cancelar</button>
      </div>
    </div>
  );
}
