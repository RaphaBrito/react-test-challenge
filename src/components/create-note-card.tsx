import { useState } from "react";

import { NoteFormCard } from "./note-form-card";

import type { NoteFormData } from "../types/note";

export interface CreateNoteCardProps {
  onCreate: (formData: NoteFormData) => void;
}

export function CreateNoteCard({ onCreate }: CreateNoteCardProps) {
  const [isCreating, setIsCreating] = useState(false);

  const handleCreate = (formData: NoteFormData) => {
    onCreate(formData);
    setIsCreating(false);
  };

  const handleCancel = () => {
    setIsCreating(false);
  };

  if (isCreating) {
    return <NoteFormCard onConfirm={handleCreate} onCancel={handleCancel} />;
  }

  return <button onClick={() => setIsCreating(true)}>+</button>;
}
