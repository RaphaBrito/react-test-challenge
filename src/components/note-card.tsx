import { useState } from "react";

import styles from "../styles/card.module.css";

import { NoteFormCard } from "./note-form-card";

import type { Note, NoteFormData } from "../types/note";

export interface NoteCardProps {
  note: Note;
  onEdit: (formData: Note) => void;
  onDelete: (id: number) => void;
}

export function NoteCard({ note, onDelete, onEdit }: NoteCardProps) {
  const [isEditing, setIsEditing] = useState(false);

  const handleBeginEditing = () => {
    setIsEditing(true);
  };

  const handleCancelEditing = () => {
    setIsEditing(false);
  };

  const handleEdit = (formData: NoteFormData) => {
    onEdit({ id: note.id, ...formData });
    setIsEditing(false);
  };

  const handleDelete = () => {
    onDelete(note.id);
  };

  if (isEditing) {
    return (
      <NoteFormCard
        note={note}
        onConfirm={handleEdit}
        onCancel={handleCancelEditing}
      />
    );
  }

  return (
    <div className={styles.card}>
      <h3>{note.title}</h3>
      <p>{note.description}</p>

      <div className={styles.cardFooter}>
        <button onClick={handleBeginEditing}>Editar</button>
        <button onClick={handleDelete}>Remover</button>
      </div>
    </div>
  );
}
