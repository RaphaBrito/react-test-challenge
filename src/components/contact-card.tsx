import { useState } from "react";

import EmailIcon from "../assets/icons/e-mail.png";
import PhoneIcon from "../assets/icons/phone.png";
import AvatarPlaceholder from "../assets/images/avatar.png";
import styles from "../styles/card.module.css";

import { ContactFormCard } from "./contact-form-card";

import type { Contact, ContactFormData } from "../types/contact";

export interface ContactCardProps {
  contact: Contact;
  onEdit: (formData: Contact) => void;
  onDelete: (id: number) => void;
}

export function ContactCard({ contact, onDelete, onEdit }: ContactCardProps) {
  const [isEditing, setIsEditing] = useState(false);

  const handleBeginEditing = () => {
    setIsEditing(true);
  };

  const handleCancelEditing = () => {
    setIsEditing(false);
  };

  const handleEdit = (formData: ContactFormData) => {
    onEdit({ id: contact.id, ...formData });
    setIsEditing(false);
  };

  const handleDelete = () => {
    onDelete(contact.id);
  };

  if (isEditing) {
    return (
      <ContactFormCard
        contact={contact}
        onConfirm={handleEdit}
        onCancel={handleCancelEditing}
      />
    );
  }

  return (
    <div className={styles.card}>
      <img className={styles.cardImage} src={AvatarPlaceholder} alt="avatar" />

      <h3>{contact.name}</h3>

      <div className={styles.cardContactInfo}>
        <img height="20px" src={PhoneIcon} alt="phone icon" />
        <p>{contact.phone}</p>
      </div>

      <div className={styles.cardContactInfo}>
        <img height="20px" src={EmailIcon} alt="e-mail icon" />
        <p>{contact.email}</p>
      </div>

      <div className={styles.cardFooter}>
        <button onClick={handleBeginEditing}>Editar</button>
        <button onClick={handleDelete}>Remover</button>
      </div>
    </div>
  );
}
