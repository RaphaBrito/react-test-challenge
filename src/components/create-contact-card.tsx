import { useState } from "react";

import { ContactFormCard } from "@/components/contact-form-card";
import { type ContactFormData } from "@/types/contact";

export interface CreateContactCardProps {
  onCreate: (formData: ContactFormData) => void;
}

export function CreateContactCard({ onCreate }: CreateContactCardProps) {
  const [isCreating, setIsCreating] = useState(false);

  const handleCreate = (formData: ContactFormData) => {
    onCreate(formData);
    setIsCreating(false);
  };

  const handleCancel = () => {
    setIsCreating(false);
  };

  if (isCreating) {
    return <ContactFormCard onConfirm={handleCreate} onCancel={handleCancel} />;
  }

  return <button onClick={() => setIsCreating(true)}>+</button>;
}
