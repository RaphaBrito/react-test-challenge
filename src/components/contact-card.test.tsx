import "@/test/utils";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";

import { ContactCard, type ContactCardProps } from "@/components/contact-card";
import { type Contact } from "@/types/contact";

describe("contact card", () => {
  it("should render correctly", () => {
    const contact = {
      id: 0,
      name: "contact name",
      email: "contact email",
      phone: "contact phone",
    } satisfies Contact;

    const props = {
      contact,
      onEdit: () => undefined,
      onDelete: () => undefined,
    } satisfies ContactCardProps;

    const { container } = render(<ContactCard {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should call onDelete", async () => {
    const onDelete = vi.fn();

    const contact = {
      id: 0,
      name: "contact name",
      email: "contact email",
      phone: "contact phone",
    } satisfies Contact;

    const props = {
      contact,
      onEdit: () => undefined,
      onDelete,
    } satisfies ContactCardProps;

    render(<ContactCard {...props} />);

    const deleteButton = screen.getByText("Remover");
    expect(deleteButton).toBeInTheDocument();

    await userEvent.click(deleteButton);
    expect(onDelete).toBeCalledWith(contact.id);
  });

  it("should call onEdit with modified data", async () => {
    const onEdit = vi.fn();

    const contact = {
      id: 0,
      name: "contact name",
      email: "contact email",
      phone: "contact phone",
    } satisfies Contact;

    const formData = contact;

    const props = {
      contact,
      onEdit,
      onDelete: () => undefined,
    } satisfies ContactCardProps;

    render(<ContactCard {...props} />);

    let editButton = screen.getByText("Editar");
    expect(editButton).toBeInTheDocument();

    await userEvent.click(editButton);
    expect(editButton).not.toBeInTheDocument();

    // Card "flipped" to editing mode
    const saveButton = screen.getByText("Salvar");
    expect(saveButton).toBeInTheDocument();

    await userEvent.click(saveButton);

    // Has returned to the viewing mode
    editButton = screen.getByText("Editar");
    expect(editButton).toBeInTheDocument();

    expect(onEdit).toBeCalledWith(formData);
  });

  it("should cancel when editing", async () => {
    const contact = {
      id: 0,
      name: "contact name",
      email: "contact email",
      phone: "contact phone",
    } satisfies Contact;

    const props = {
      contact,
      onEdit: () => undefined,
      onDelete: () => undefined,
    } satisfies ContactCardProps;

    render(<ContactCard {...props} />);

    let editButton = screen.getByText("Editar");
    expect(editButton).toBeInTheDocument();

    await userEvent.click(editButton);
    expect(editButton).not.toBeInTheDocument();

    // Card "flipped" to editing mode
    const cancelButton = screen.getByText("Cancelar");
    expect(cancelButton).toBeInTheDocument();

    await userEvent.click(cancelButton);

    // Has returned to the viewing mode
    editButton = screen.getByText("Editar");
    expect(editButton).toBeInTheDocument();
  });
});
