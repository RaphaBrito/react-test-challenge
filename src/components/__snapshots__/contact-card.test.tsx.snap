// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`contact card > should render correctly 1`] = `
<div>
  <div
    class="_card_277467"
  >
    <img
      alt="avatar"
      class="_cardImage_277467"
      src="/src/assets/images/avatar.png"
    />
    <h3>
      contact name
    </h3>
    <div
      class="_cardContactInfo_277467"
    >
      <img
        alt="phone icon"
        height="20px"
        src="/src/assets/icons/phone.png"
      />
      <p>
        contact phone
      </p>
    </div>
    <div
      class="_cardContactInfo_277467"
    >
      <img
        alt="e-mail icon"
        height="20px"
        src="/src/assets/icons/e-mail.png"
      />
      <p>
        contact email
      </p>
    </div>
    <div
      class="_cardFooter_277467"
    >
      <button>
        Editar
      </button>
      <button>
        Remover
      </button>
    </div>
  </div>
</div>
`;
